rllib/examples/learners/classes/curiosity_ppo_torch_learner.py

```diff
@@ -0,0 +1,141 @@
+        if curiosity_beta is not NotProvided:
+            self.curiosity_beta = curiosity_beta
+        if curiosity_eta is not NotProvided:
+            self.curiosity_eta = curiosity_eta
+        return self
+
+
+class PPOTorchLearnerWithCuriosity(PPOTorchLearner):
+    def build(self):
+        super().build()
+
+        # Assert, we are only training one policy (RLModule).
+        assert len(self.module) == 1 and DEFAULT_MODULE_ID in self.module
+
+        # Add an InverseDynamicsModel to our MARLModule.
+        icm_spec = RLModuleSpec(
+            module_class=InverseDynamicsModel,
+            observation_space=self.module[DEFAULT_MODULE_ID].config.observation_space,
+            action_space=self.module[DEFAULT_MODULE_ID].config.action_space,
+            model_config_dict={
+                "feature_net_hiddens": self.config.curiosity_feature_net_hiddens,
+                "feature_net_activation": self.config.curiosity_feature_net_activation,
+                "inverse_net_hiddens": self.config.curiosity_inverse_net_hiddens,
+                "inverse_net_activation": self.config.curiosity_inverse_net_activation,
+                "forward_net_hiddens": self.config.curiosity_forward_net_hiddens,
+                "forward_net_activation": self.config.curiosity_forward_net_activation,
+            },
+        )
+        self.add_module(
+            module_id=ICM_MODULE_ID,
+            module_spec=icm_spec,
+        )
+
+        # Prepend a "add-NEXT_OBS-from-episodes-to-train-batch" connector piece (right
+        # after the corresponding "add-OBS-..." default piece).
+        if self.config.add_default_connectors_to_learner_pipeline:
+            self._learner_connector.insert_after(
+                AddObservationsFromEpisodesToBatch,
+                AddNextObservationsFromEpisodesToTrainBatch(),
+            )
+
+    def compute_loss(
+        self,
+        *,
+        fwd_out: Dict[str, Any],
+        batch: Dict[str, Any],
+    ) -> Dict[str, Any]:
+        # Compute the ICM loss first (so we'll have the chance to change the rewards
+        # in the batch for the "main" RLModule (before we compute its loss with the
+        # intrinsic rewards).
+        icm = self.module[ICM_MODULE_ID]
+        # Send the exact same batch to the ICM module that we used for the "main"
+        # RLModule's forward pass.
+        icm_fwd_out = icm.forward_train(batch=batch[DEFAULT_MODULE_ID])
+        # Compute the loss of the ICM module.
+        icm_loss = icm.compute_loss_for_module(
+            learner=self,
+            module_id=ICM_MODULE_ID,
+            config=self.config.get_config_for_module(ICM_MODULE_ID),
+            batch=batch[DEFAULT_MODULE_ID],
+            fwd_out=icm_fwd_out,
+        )
+
+        # Add intrinsic rewards from ICM's `fwd_out` (multiplied by factor `eta`)
+        # to "main" module batch's extrinsic rewards.
+        batch[DEFAULT_MODULE_ID][Columns.REWARDS] += (
```

Comment:
We could also write a connector piece to perform this logic, couldn't we?

Reply:
Great question. We might be able to do this, but only if we make 100% sure that the ICM loss (intrinsic rewards are the individual per-batch-item forward loss values) if available before(!) we call the main RLModule's `compute_loss_for_module` method.

For this to work, we'd probably make the loss computation of the ICM already in its `forward_train` call, which is kind of counterintuitive (the `forward_train` should probably NOT compute any actual losses yet). Nut not sure, this is more a philosophical discussion.
So yes, it would probably work also with a custom connector (instead of overriding `compute_loss`).